codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java

```diff
@@ -379,21 +379,36 @@ public T add(Headers<? extends K, ? extends V, ?> headers) {
         if (headers == this) {
             throw new IllegalArgumentException("can't add to itself.");
         }
+        addImpl(headers);
+        return thisT();
+    }
+
+    protected void addImpl(Headers<? extends K, ? extends V, ?> headers) {
         if (headers instanceof DefaultHeaders) {
             @SuppressWarnings("unchecked")
-            DefaultHeaders<K, V, T> defaultHeaders = (DefaultHeaders<K, V, T>) headers;
-            HeaderEntry<K, V> e = defaultHeaders.head.after;
-            while (e != defaultHeaders.head) {
-                add(e.key, e.value);
-                e = e.after;
+            DefaultHeaders<? extends K, ? extends V, T> defaultHeaders =
+                    (DefaultHeaders<? extends K, ? extends V, T>) headers;
+            HeaderEntry<? extends K, ? extends V> e = defaultHeaders.head.after;
+            if (defaultHeaders.hashingStrategy == this.hashingStrategy &&
+                    defaultHeaders.nameValidator == this.nameValidator) {
+                // Fastest copy
+                while (e != defaultHeaders.head) {
+                    add0(e.hash, index(e.hash), e.key, e.value);
+                    e = e.after;
+                }
+            } else {
+                // Fast copy
```

Comment:
Here's some for 'fastest' vs 'fast'. Adding benchmark to PR

Benchmark                         (exampleHeader)  Mode  Cnt    Score    Error  Units
HeadersBenchmark.http2AddAllFastest           SIX  avgt   20  149.917 ±  9.492  ns/op
HeadersBenchmark.http2AddAllFast              SIX  avgt   20  246.499 ± 18.825  ns/op
HeadersBenchmark.httpAddAllFastest               SIX  avgt   20  109.263 ± 13.840  ns/op